### ISSUE TITLE
aws_ec2 example fails

### ISSUE THREAD

**Issue by treaves** [rancher/terraform-provider-rke#186]
I'm trying the aws_ec2 example from this repo.  The terraform apply fails with the following:
```
module.nodes.aws_instance.rke-node[2]: Creation complete after 1m28s [id=i-0e960c32131c8242e]

Error: error executing "/tmp/terraform_1474988253.sh": Process exited with status 6

Error: error executing "/tmp/terraform_1748626233.sh": Process exited with status 6
```

In looking at the aws console, all four instances are created and running.  It seems the scripts are ending early, and therefore the `kube_config_cluster.yml ` isn't being created.  I can't find any timeout setting to adjust.  Am I missing something?

Thanks.

**Comment by srirajan** [comment]
I can see the same issue. A weird quirk is that if I keep running this, I do get some success. I have not tested this but it seems like only 1 instance succeeds.  So, if I have 6 instances, i get 5 failures. Run it again, and, I get 4 failures and so on. Any tips?

**Comment by srirajan** [comment]
An update after more testing. This seems to be a persistent problem.  The cluster is provisioned correctly as I can manually create kubeconfig from the Terraform output and it responds fine.  The problem seems to be with the remote provisioner. I also tried the 19.03 version of the script and got the same issue.

Here's my template

```
# Get AMI
data "aws_ami" "rke_os" {
  most_recent = true

  filter {
    name   = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-xenial-16.04-amd64-server-*"]
  }

  filter {
    name   = "virtualization-type"
    values = ["hvm"]
  }

  owners = ["099720109477"] # Canonical
}


# Step 1: Create an IAM role
resource "aws_iam_role" "rke_role" {
  name = "rancher-rke-role"

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": {
    "Effect": "Allow",
    "Principal": {"Service": "ec2.amazonaws.com"},
    "Action": "sts:AssumeRole"
  }
}
EOF
}

# Step 2: Add  Access Policy
resource "aws_iam_role_policy" "rke-access-policy" {
  name = "rancher-rke-access-policy"
  role = aws_iam_role.rke_role.id

  policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "ec2:Describe*",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "ec2:AttachVolume",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": "ec2:DetachVolume",
      "Resource": "*"
    },
    {
      "Effect": "Allow",
      "Action": ["ec2:*"],
      "Resource": ["*"]
    },
    {
      "Effect": "Allow",
      "Action": ["elasticloadbalancing:*"],
      "Resource": ["*"]
    }
  ]
}
EOF
}

# Step 3: Create the Instance Profile
resource "aws_iam_instance_profile" "rke_aws" {
  name = "rancher-rke"
  role = aws_iam_role.rke_role.name
}

resource "tls_private_key" "rke_node_ssh_key" {
  algorithm = "RSA"
}

resource "aws_key_pair" "rke_node_ssh_key_pair" {
  key_name   = "rke-node-key"
  public_key = tls_private_key.rke_node_ssh_key.public_key_openssh
}

locals {
  cluster_id_tag = {
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
  }
}
resource "aws_security_group" "rke_sg" {
  name        = "rke-default-security-group"
  description = "rke"
  vpc_id                 = module.vpc.vpc_id

  ingress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }

  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"

  }

}

resource "aws_instance" rke_controlplane_az1 {
  count                  = 2
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[0]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "controlplane"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }

  user_data = templatefile("./userdata_rancher_server.template", {
    rke_docker_version = var.rke_docker_version
    rke_ssh_username       = var.rke_ssh_username
  })

  provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = self.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
}

resource "aws_instance" rke_controlplane_az2 {
  count                  = 1
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[1]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "controlplane"

  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }

  user_data = templatefile("./userdata_rancher_server.template", {
    rke_docker_version = var.rke_docker_version
    rke_ssh_username       = var.rke_ssh_username
  })

  provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = self.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
}

resource "aws_instance" rke_worker_az1 {
  count                  = 1
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[0]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "worker"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }

  user_data = templatefile("./userdata_rancher_server.template", {
    rke_docker_version = var.rke_docker_version
    rke_ssh_username       = var.rke_ssh_username
  })

  provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = self.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
}


resource "aws_instance" rke_worker_az2 {
  count                  = 1
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[1]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "worker"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }

  user_data = templatefile("./userdata_rancher_server.template", {
    rke_docker_version = var.rke_docker_version
    rke_ssh_username       = var.rke_ssh_username
  })

  provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = self.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
}

resource "rke_cluster" "cluster" {
  cloud_provider {
    name = "aws"
  }

  # 3 control plane nodes in 2 AZs
  nodes {
    address          = aws_instance.rke_controlplane_az1[0].public_dns
    internal_address = aws_instance.rke_controlplane_az1[0].private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["controlplane", "etcd"]
  }

  nodes {
    address          = aws_instance.rke_controlplane_az1[1].public_dns
    internal_address = aws_instance.rke_controlplane_az1[1].private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["controlplane", "etcd"]
  }

  nodes {
    address          = aws_instance.rke_controlplane_az2[0].public_dns
    internal_address = aws_instance.rke_controlplane_az2[0].private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["controlplane", "etcd"]
  }

  # 3 worker nodes in 2 AZs
  nodes {
    address          = aws_instance.rke_worker_az1[0].public_dns
    internal_address = aws_instance.rke_worker_az1[0].private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["worker"]
  }

  nodes {
    address          = aws_instance.rke_worker_az2[0].public_dns
    internal_address = aws_instance.rke_worker_az2[0].private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["worker"]
  }
}

resource "local_file" "kube_cluster_yaml" {
  filename = "./kube_config_cluster.yml"
  content  = rke_cluster.cluster.kube_config_yaml
}
```

**Comment by srirajan** [comment]
Adding a null resource and some dependencies seem to help

```
resource "aws_instance" rke_controlplane_node1 {
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[0]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "controlplane"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }
}

resource "null_resource" "rke_controlplane_node1" {
  connection {
      host        = aws_instance.rke_controlplane_node1.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
  }

    provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = aws_instance.rke_controlplane_node1.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
  depends_on = [aws_instance.rke_controlplane_node1]
}

resource "aws_instance" rke_controlplane_node2 {
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[0]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "controlplane"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }
}

resource "null_resource" "rke_controlplane_node2" {
  connection {
      host        = aws_instance.rke_controlplane_node2.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
  }

    provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = aws_instance.rke_controlplane_node2.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
  depends_on = [aws_instance.rke_controlplane_node2]
}

resource "aws_instance" rke_worker_node1 {
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[0]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "worker"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }
}

resource "null_resource" "rke_worker_node1" {
  connection {
      host        = aws_instance.rke_worker_node1.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
  }

    provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = aws_instance.rke_worker_node1.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
  depends_on = [aws_instance.rke_worker_node1]
}

resource "aws_instance" rke_worker_node2 {
  instance_type          = var.rke_cp_instance_type
  ami                    = data.aws_ami.rke_os.id
  key_name               = aws_key_pair.rke_node_ssh_key_pair.id
  ebs_optimized          = true
  vpc_security_group_ids = [aws_security_group.rke_sg.id]
  subnet_id              = module.vpc.public_subnets[1]

  tags = {
    project                                       = var.project_name
    creator                                       = var.creator
    "kubernetes.io/cluster/${var.rke_cluster_id}" = "owned"
    k8s_role                                      = "worker"
  }

  root_block_device {
    volume_type = var.primary_ebs_volume_type
    volume_size = var.primary_ebs_volume_size
  }
}

resource "null_resource" "rke_worker_node2" {
  connection {
      host        = aws_instance.rke_worker_node2.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
  }

    provisioner "remote-exec" {
    connection {
      type        = "ssh"
      host        = aws_instance.rke_worker_node2.public_ip
      user        = var.rke_ssh_username
      private_key = tls_private_key.rke_node_ssh_key.private_key_pem
    }

    inline = [
      "curl ${var.rke_docker_install_url} | sh",
      "sudo usermod -a -G docker ubuntu",
    ]
  }
  depends_on = [aws_instance.rke_worker_node2]
}

resource "rke_cluster" "cluster" {
  # cloud_provider {
  #   name = "aws"
  # }

  # 2 control plane nodes in 2 AZs
  nodes {
    address          = aws_instance.rke_controlplane_node1.public_dns
    internal_address = aws_instance.rke_controlplane_node1.private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["controlplane", "etcd"]
  }

  nodes {
    address          = aws_instance.rke_controlplane_node2.public_dns
    internal_address = aws_instance.rke_controlplane_node2.private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["controlplane", "etcd"]
  }
  
  # 2 worker nodes in 2 AZs
  nodes {
    address          = aws_instance.rke_worker_node1.public_dns
    internal_address = aws_instance.rke_worker_node1.private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["worker"]
  }

  nodes {
    address          = aws_instance.rke_worker_node2.public_dns
    internal_address = aws_instance.rke_worker_node2.private_ip
    user             = var.rke_ssh_username
    ssh_key          = tls_private_key.rke_node_ssh_key.private_key_pem
    role             = ["worker"]
  }

  depends_on = [null_resource.rke_controlplane_node1, null_resource.rke_controlplane_node2, null_resource.rke_worker_node1, null_resource.rke_worker_node2]

}

resource "local_file" "kube_cluster_yaml" {
  filename = "./kube_config_cluster.yml"
  content  = rke_cluster.cluster.kube_config_yaml
}
```

**Comment by nikkelma** [comment]
Without knowing exactly what `userdata_rancher_server.template` contains, but having a guess it could be similar to [this file from the quickstart](https://github.com/rancher/quickstart/blob/3b25a4bd0030c6b895881a06af65280ba9c2e03e/cloud-common/files/userdata_rancher_server.template), I'd say a likely root cause is installing docker twice on each node. In the first set of code, the userdata script runs on first boot, and then your remote provisioner was trying to run the same (or very similar) install script and failing.

I believe the null resource changes were a bit of a red herring, as it seems you also removed the userdata sections from the nodes as well, meaning docker is now only being installed once per node. I'd be curious to see if you could use a configuration similar to the first one but choosing to install either through userdata or through the remote provisioner.

**Comment by srirajan** [comment]
It is the file from the quickstart. The first set of code (inline) failed repeatedly. It also re-provisions the ec2 instance and so it can't be running it twice. I will give your idea a try to use the first set of code and use remote provisioner. Thanks

**Comment by rawmind0** [comment]
Please, reopen issue if needed.

**Comment by AjdinDev** [comment]
@srirajan @treaves Did you ever end up finding the cause behind the error? Running into the same issue right now.

**Comment by treaves** [comment]
@AjdinDev No; I just went in another direction instead.

**Comment by AjdinDev** [comment]
Then I think it'd be a good idea to reopen this issue @rawmind0.

I'm having the same issue as above, first node is successfully getting docker setup and the rest are failing. Using default AWS example with the below error:

`module.nodes.aws_instance.rke-node[4] (remote-exec): Server: Docker Engine - Community
module.nodes.aws_instance.rke-node[4] (remote-exec):  Engine:
module.nodes.aws_instance.rke-node[4] (remote-exec):   Version:          18.09.9
module.nodes.aws_instance.rke-node[4] (remote-exec):   API version:      1.39 (minimum version 1.12)
module.nodes.aws_instance.rke-node[4] (remote-exec):   Go version:       go1.11.13
module.nodes.aws_instance.rke-node[4] (remote-exec):   Git commit:       039a7df
module.nodes.aws_instance.rke-node[4] (remote-exec):   Built:            Wed Sep  4 16:19:38 2019
module.nodes.aws_instance.rke-node[4] (remote-exec):   OS/Arch:          linux/amd64
module.nodes.aws_instance.rke-node[4] (remote-exec):   Experimental:     false

module.nodes.aws_instance.rke-node[4] (remote-exec): If you would like to use Docker as a non-root user, you should now consider
module.nodes.aws_instance.rke-node[4] (remote-exec): adding your user to the "docker" group with something like:

module.nodes.aws_instance.rke-node[4] (remote-exec):   sudo usermod -aG docker ubuntu

module.nodes.aws_instance.rke-node[4] (remote-exec): Remember that you will have to log out and back in for this to take effect!

module.nodes.aws_instance.rke-node[4] (remote-exec): WARNING: Adding a user to the "docker" group will grant the ability to run
module.nodes.aws_instance.rke-node[4] (remote-exec):          containers which can be used to obtain root privileges on the
module.nodes.aws_instance.rke-node[4] (remote-exec):          docker host.
module.nodes.aws_instance.rke-node[4] (remote-exec):          Refer to https://docs.docker.com/engine/security/security/#docker-daemon-attack-surface
module.nodes.aws_instance.rke-node[4] (remote-exec):          for more information.

module.nodes.aws_instance.rke-node[4]: Creation complete after 1m40s [id=i-082e01f97d17e76b7]

Error: error executing "/tmp/terraform_1943988503.sh": Process exited with status 6



Error: error executing "/tmp/terraform_1108958112.sh": Process exited with status 6



Error: error executing "/tmp/terraform_125852443.sh": Process exited with status 6



Error: error executing "/tmp/terraform_84720841.sh": Process exited with status 6`

**Comment by srirajan** [comment]
I just bootstrapped my own instance.  Started with a vanilla Ubuntu AMI and used https://releases.rancher.com/install-docker/18.09.sh to install Docker via userdata. I have not used this in a while or on a lot of systems and so can't vouch if that solves for all use cases. In my case, I did get success on a few tries of installing everything from scratch.

**Comment by AjdinDev** [comment]
After some debugging, I managed to resolve the issue.

Steps I took:

- Used Ubuntu 18.04 instead of 16.x (default)
- Update the rancher docker install script to `https://releases.rancher.com/install-docker/19.03.sh` 

Both of these seemed to have fixed the issue. I believe that rancher was having issues finding a good docker version for ubuntu 16.

I might make a PR to "update" the AWS example to more recent versions of Ubuntu and the install script.
